packages/wallet-sdk/src/core/communicator/Communicator.ts

```diff
@@ -96,5 +110,35 @@ export class Communicator {
         if (!this.popup) throw standardErrors.rpc.internal();
         return this.popup;
       });
-  }
+  };
+
+  /**
+   * Posts a RPC request to the popup window and waits for a response
+   */
+  postRPCRequest = (request: RPCRequestMessage): Promise<RPCResponseMessage> => {
+    this.pendingRequests.set(request.id, 'created');
+
+    return new Promise((resolve, reject) => {
+      // reject if the request was cancelled between creation and awaiting due to disconnect
+      if (!this.pendingRequests.has(request.id)) {
+        reject(standardErrors.provider.userRejectedRequest('Request cancelled before sending'));
+        return;
+      }
```

Comment:
non-`awaiting` requests can be cancelled without getting posted to popup, if popup explicitly has sent `PopupUnload` event before it's awaited.